File: 9.BinarySearch/C/main.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#define MAX_STR_LEN 9
#define MAX_STR_COUNT 10000
#define MAX_INT_COUNT 10000

void loadData(int n_lines, char strings[][MAX_STR_LEN], int ints[]) {
    FILE *f;
    int i = 0, j = 0;
    char line[100];

    f = fopen("../strings.txt", "r");
    while (i < n_lines && fgets(line, sizeof(line), f)) {
        line[strcspn(line, "\n")] = 0;  // remove newline
        strncpy(strings[i], line, MAX_STR_LEN);
        strings[i++][MAX_STR_LEN] = '\0';
    }
    fclose(f);

    f = fopen("../ints.txt", "r");
    while (j < n_lines && fgets(line, sizeof(line), f)) {
        ints[j++] = atoi(line);
    }
    fclose(f);
}

int cmpstr(const void* a, const void* b) {
    return strcmp((char*)a, (char*)b);
}

int cmpint(const void* a, const void* b) {
    return (*(int*)a - *(int*)b);
}

int main() {
    int n_lines = 5000;
    char strings[n_lines][MAX_STR_LEN];
    int ints[n_lines];
    loadData(n_lines, strings, ints);

    char* searchString = strings[0];
    int searchInt = ints[0];

    clock_t start, end;
    double cpu_time_used;

    start = clock();
    qsort(strings, n_lines, MAX_STR_LEN, cmpstr);
    char* resultString = (char*)bsearch(searchString, strings, n_lines, MAX_STR_LEN, cmpstr);
    end = clock();
    cpu_time_used = ((double) (end - start)) / CLOCKS_PER_SEC;
    printf("Target string %s found: %s\n", searchString, resultString);
    printf("String lookup time: %.6f seconds.\n", cpu_time_used);

    start = clock();
    qsort(ints, n_lines, sizeof(int), cmpint);
    int* resultInt = (int*)bsearch(&searchInt, ints, n_lines, sizeof(int), cmpint);
    end = clock();
    cpu_time_used = ((double) (end - start)) / CLOCKS_PER_SEC;
    printf("Target integer %d found: %d\n", searchInt, *resultInt);
    printf("Integer lookup time: %.6f seconds.\n", cpu_time_used);

    return 0;
}

File: 9.BinarySearch/Go/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
	"strconv"
	"time"
)

func loadData(n_lines int, strings *[]string, ints *[]int) {
    file, _ := os.Open("../strings.txt")
    scanner := bufio.NewScanner(file)
    for i := 0; scanner.Scan() && i < n_lines; i++ {
        *strings = append(*strings, scanner.Text())
    }
    file.Close()

    file, _ = os.Open("../ints.txt")
    scanner = bufio.NewScanner(file)
    for i := 0; scanner.Scan() && i < n_lines; i++ {
        num, _ := strconv.Atoi(scanner.Text())
        *ints = append(*ints, num)
    }
    file.Close()
}

func main() {
    n_lines := 5000
    var strings []string
    var ints []int
    loadData(n_lines, &strings, &ints)

    searchString := strings[0]
    searchInt := ints[0]

	start := time.Now()
    sort.Strings(strings)
    resultString := sort.SearchStrings(strings, searchString)
	elapsed := time.Since(start)
    fmt.Printf("Target string %s found: %s\n", searchString, strings[resultString]);
    fmt.Printf("String lookup time: %.6f seconds.\n", elapsed.Seconds());

	start = time.Now()
    sort.Ints(ints)
    resultInt := sort.SearchInts(ints, searchInt)
	elapsed = time.Since(start)
    fmt.Printf("Target int %d found: %d\n", searchInt, ints[resultInt]);
    fmt.Printf("Int lookup time: %.6f seconds.\n", elapsed.Seconds());
}
